File: WordPattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<char, string> charToWord;
        unordered_set<string> usedWords;
        vector<string> strings;
        string current = "";
        for(char c : s){
            if(c != ' '){
                current += c;
            } else {
                strings.push_back(current);
                current = "";
            }
        }
        if(!current.empty()){
            strings.push_back(current);
        }
        if(pattern.length() != strings.size()){
            return false;
        }
        for (int i = 0; i < pattern.size(); ++i) {
            char c = pattern[i];
            string word = strings[i];

            if (charToWord.count(c)) {
                if (charToWord[c] != word) {
                    return false; 
                }
            } else {
                if (usedWords.count(word)) {
                    return false;
                }
                charToWord[c] = word;
                usedWords.insert(word);
            }
        }

        return true;
    }
};

File: WordPattern.ts
function wordPattern(pattern: string, s: string): boolean {
    const words: string[] = s.split(' ');
    if (pattern.length !== words.length) {
        return false;
    }
    let wordToChar: Map<string, string> = new Map();
    let charToWord: Map<string, string> = new Map();

    for(let i = 0; i < pattern.length; ++i){
        const c = pattern[i];
        const w = words[i];
        if (wordToChar.has(w)) {
            if (wordToChar.get(w) != c) {
                return false;
            }
        } else {
            wordToChar.set(w, c);
        }
        if (charToWord.has(c)) {
            if (charToWord.get(c) != w) {
                return false;
            }
        } else {
            charToWord.set(c, w);
        }
    }
    return true;
};
